File: src/main/java/com/kisman/cc/gui/api/Component.java
package com.kisman.cc.gui.api;

import com.kisman.cc.gui.halq.HalqGui;

public interface Component {
    default void drawScreen(int mouseX, int mouseY) { }
    default void drawScreenPost(int mouseX, int mouseY) { HalqGui.renderComponent(this); }
    default void mouseClicked(int mouseX, int mouseY, int button) { }
    default void mouseReleased(int mouseX, int mouseY, int mouseButton) { }
    default void updateComponent(int x, int y) { }
    default void keyTyped(char typedChar, int key) { }
    default void setOff(int newOff) { }
    default void setCount(int count) { }
    default int getHeight() { return getRawHeight(); }
    default int getRawHeight() { return HalqGui.height; }
    default int getCount() { return 0; }
    default void setWidth(int width) { }
    default void setX(int x) { }
    default int getX() { return 0; }
    default void setY(int y) { }
    default int getY() { return 0; }
    default void setLayer(int layer) { }
    default int getLayer() { return 0; }
    default boolean visible() {return true;}
}

File: src/main/java/com/kisman/cc/gui/api/Component.kt
package com.kisman.cc.gui.api

import com.kisman.cc.gui.halq.HalqGui
import com.kisman.cc.gui.halq.components.Description

@Suppress("UNUSED_PARAMETER")
interface Component {
    fun drawScreen(
        mouseX : Int,
        mouseY : Int
    ) {
        HalqGui.currentComponent = this
    }

    fun mouseClicked(mouseX : Int, mouseY :  Int, button : Int) { }
    fun mouseReleased(mouseX : Int, mouseY : Int, mouseButton : Int) { }
    fun updateComponent(x : Int, y : Int) { }
    fun keyTyped(typedChar : Char, key : Int) { }
    fun setOff(newOff : Int) { }

    val height : Int
        get() = HalqGui.height
    val rawHeight : Int
        get() = HalqGui.height
    val width : Int
    var count : Int

    var x : Int
        get() = 0
        set(value) { }

    var y : Int
        get() = 0
        set(value) { }

    var layer : Int
        get() = 0
        set(value) { }

    fun visible() : Boolean = true

    fun isMouseOnButton(
        x : Int,
        y : Int
    ) : Boolean = false

    var description : Description?
        get() = null
        set(value) { }
}